### PR TITLE
feat(cdk/tree): allow use of nested node with flattened data, and vice versa

### PR DESCRIPTION
**Pull request by BobobUnicorn**
Based on #25729. Followed by #26585.

### CODE REVIEW

File: src/components-examples/cdk/tree/cdk-tree-flat-children-accessor/cdk-tree-flat-children-accessor-example.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  templateUrl: 'cdk-tree-flat-children-accessor-example.html',
+  styleUrls: ['cdk-tree-flat-children-accessor-example.css'],
+})
+export class CdkTreeFlatChildrenAccessorExample {
+  @ViewChild(CdkTree)
+  tree!: CdkTree<NestedFoodNode>;
+
+  childrenAccessor = (dataNode: NestedFoodNode) => timer(100).pipe(mapTo(dataNode.children ?? []));
+
+  dataSource = new ArrayDataSource(NESTED_DATA);
```
Comment (richieforeman) [inline src/components-examples/cdk/tree/cdk-tree-flat-children-accessor/cdk-tree-flat-children-accessor-example.ts:33]: Nit: Can be readonly?

File: src/components-examples/cdk/tree/cdk-tree-nested-level-accessor/cdk-tree-nested-level-accessor-example.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ */
+@Component({
+  selector: 'cdk-tree-nested-level-accessor-example',
+  templateUrl: 'cdk-tree-nested-level-accessor-example.html',
+  styleUrls: ['cdk-tree-nested-level-accessor-example.css'],
+})
+export class CdkTreeNestedLevelAccessorExample {
+  levelAccessor = (dataNode: FlatFoodNode) => dataNode.level;
+
+  dataSource = new ArrayDataSource(FLAT_DATA);
```
Comment (richieforeman) [inline src/components-examples/cdk/tree/cdk-tree-nested-level-accessor/cdk-tree-nested-level-accessor-example.ts:16]: Nit: can be readonly?

File: src/cdk/tree/tree.ts [hunk 3]
```diff
@@ -146,6 +149,16 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
    */
   @Input() expansionKey?: (dataNode: T) => K;
 
+  /**
+   * What type of node is being used in the tree. This must be provided if either of
+   * `levelAccessor` or `childrenAccessor` are provided.
+   *
+   * This controls what selection of data the tree will render.
+   */
+  // NB: we're unable to determine this ourselves; Angular's ContentChildren
```
Comment (jelbourn) [inline src/cdk/tree/tree.ts:158]: "NB"?

File: src/cdk/tree/tree.ts [hunk 4]
```diff
@@ -146,6 +149,16 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
...
+  // unfortunately does not pick up the necessary information.
+  @Input() nodeType?: 'flat' | 'nested';
```
Comment (jelbourn) [inline src/cdk/tree/tree.ts:160]: Why is this necessary at the node level? I would think that this is only needed at the entire tree level

File: src/cdk/tree/tree.ts [hunk 5]
```diff
@@ -265,21 +296,23 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
 
     if (dataStream) {
       this._dataSubscription = dataStream
-        .pipe(takeUntil(this._onDestroy))
-        .subscribe(data => this.renderNodeChanges(data));
+        .pipe(
+          switchMap(data => this._convertChildren(data)),
```
Comment (jelbourn) [inline src/cdk/tree/tree.ts:300]: I don't quite follow what's going on with this conversation here

File: src/cdk/tree/tree.ts [hunk 6]
```diff
@@ -570,6 +679,39 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
...
+
+  /**
+   * Converts children for certain tree configurations. Note also that this
+   * caches the known nodes for use in other parts of the tree.
+   */
+  private _convertChildren(nodes: readonly T[]): Observable<readonly T[]> {
+    // The only situations where we have to convert children types is when
+    // they're mismatched; i.e. if the tree is using a childrenAccessor and the
+    // nodes are flat, or if the tree is using a levelAccessor and the nodes are
+    // nested.
```
Comment (jelbourn) [inline src/cdk/tree/tree.ts:691]: I don't quite follow how this situation would come about- I would think we would only want to support _either_ flat or nested for the entire free instance
